core/src/feature/mod.rs

```diff
@@ -8,42 +8,39 @@ use crate::feature::descriptor::FeatureDescriptor;
 use crate::uri::UriBound;
 
 /// Represents extension data for a given feature.
-/// # Unsafety
-/// Since extension data is passed to plugin as a raw pointer,
-/// structs implementing this trait must be `#[repr(C)]`.
-pub unsafe trait Feature: Sized + Copy {
-    const URI: &'static [u8];
-
+pub trait Feature: Sized + Copy + UriBound {
```

Comment:
Probably a copy/paste mistake on your end? This trait should still be `unsafe` to implement, as because of the `#[repr(C)]` requirement the struct will end up getting `transmute`'d from the host in C.

Reply:
No, this is intentional: First of all, being `#[repr(C)]` doesn't make something unsafe, it's really just another way to represent data in Rust. Also, everything that is important to make it usable in C is already guarded by the super-traits: Features have to be sized and they have to be easily copyable. What other requirements/promises must be kept by a feature that can't be expressed with super-traits?

Reply:
I agree it doesn't make it unsafe, but the way I see it is the other way around: we cannot enforce that a struct is indeed `#[repr(C)]` through traits, so the developer of the feature (which may not be us) has to make sure it is indeed marked `#[repr(C)]`, otherwise the ABI is not guaranteed at all to be correct: Rust may reorder fields to optimize alignment, and some Rust-specific types (thick pointers like `&str`, or enums, both of which are `Copy` and `Sized`) must not be used as they have no ABI defined for the C end.

The `unsafe` is also here to represent the contract that the struct has to have *exactly* the same ABI as the one defined in the corresponding LV2 spec, and that the associated URL is the exact same. If any of those two conditions are not met, the whole `Feature` wrapper falls apart and the plugin will go straight into undefined behavior. :/
And while `bindgen` is a great help to not make these mistakes (since both the structure and URL are auto-generated from the official header file), I don't think  the safety should be taken for granted, especially considering that other developers may create their own features outside of this lib.

Reply:
Okay. If we abstract this view, this would mean that implementing additions for other specifications is an unsafe operation. This would also imply other messages to users: a) you should not do it, if you don't need to do it. b) If you want to implement new specifications, you have to know what you are doing, since you're talking a lot with C.

I like these stances. Maybe we should add that to our code style/contribution guide: Our API should be safe to use, but unsafe to extend.

Reply:
However, this stance will lead to wide-spread, pervasive use of unsafe traits, which something the Rust community does not like to see. Quoting [the nomicon](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html):
> The decision of whether to mark a trait unsafe is an API design choice. Rust has traditionally avoided doing this because it makes Unsafe Rust pervasive, which isn't desirable. 
Do we really want this?

Reply:
> I like these stances. Maybe we should add that to our code style/contribution guide: Our API should be safe to use, but unsafe to extend.

I don't quite see it that way actually: if someone implements a new Feature (or an Extension, or a Port Type, or an Atom Type…), they are not extending _our_ API. They are extending the LV2 specification, or rather, making additional LV2 bindings to another spec, both of which are, in the end, C ABIs, and by definition are unsafe to interact with.

Therefore I don't think this really is a style/contribution guideline. However I think we should make a proper, separate documentation page explaining how to implement additional Features, what is the LV2 design behind it, and why it is unsafe to do in Rust, using some of the arguments from this discussion, and maybe linking back to it. (Hey, future you who came here from the docs! :wink: )

> However, this stance will lead to wide-spread, pervasive use of unsafe traits, which something the Rust community does not like to see. Quoting [the nomicon](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html):
> 
> > The decision of whether to mark a trait unsafe is an API design choice. Rust has traditionally avoided doing this because it makes Unsafe Rust pervasive, which isn't desirable.
> 
> Do we really want this?

I do not think the traits like `Feature` are going to be used at all by the vast majority of users.

Just like `Send` or `Sync` they are marker traits that practically never used directly: in this library there is [only one function](https://github.com/rust-dsp/rust-lv2/blob/25dd4f2f26a93254f50623fcfada2628ea7adc3f/core/src/feature/descriptor.rs#L71) were there are used for feature resolution, and even the call to that function is abstracted and derive'd away.
(It is still possible to call this method by hand if you ever need to of course, but it is not unsafe to do so, so no issue here.)

And in terms of implementing the trait, only the developers who make bindings to specs will have to `unsafe`ly implement it, which is a tiny minority. Especially if that binding to a new spec will end up in a new crate to be reused. Just like `Send` or `Sync` again, which are only manually implemented by niche data structure or bindings (and the fact that I messed up their implementation in the other PR shows the usefulness of their unsafety :slightly_smiling_face: ).

Quoting the nomicon again, I think here we fall in this very specific case:

>  Send and Sync are marked unsafe because thread safety is a fundamental property that unsafe code can't possibly hope to defend against in the way it could defend against a buggy Ord implementation.

On this level `Feature` is much closer to `Send` or `Sync` than `Ord`: there is no way we can possibly detect a faulty `Feature` implementation, because it will probably derail into jumping at some random place in memory before it even reaches Rust code again (if ever!). Messing it up is clearly ends up into a memory error, not a logic error.

On a shorter note though, I usually rely on this line from the nomicon:

> **No matter what, Safe Rust can't cause Undefined Behavior.**

If I can trigger some wild UB by just changing two letters in the URL (like `map` vs. `unmap`), then yeah, this should better be marked `unsafe` I think. :slightly_smiling_face: 

(The changes you just made look good btw, you can resolve this conversation and merge it if you just agree with what I said here.)